js-pkg/client/src/provider.ts

```diff
@@ -210,443 +50,392 @@ type WebSocketPolyfillType = {
+
+    this.websocket?.send(encoding.toUint8Array(encoder))
+  }
+
+  private updateAwareness(decoder: decoding.Decoder) {
+    awarenessProtocol.applyAwarenessUpdate(
+      this.awareness,
+      decoding.readVarUint8Array(decoder),
+      this,
+    )
+  }
+
+  private receiveSyncMessage(decoder: decoding.Decoder) {
+    const encoder = encoding.createEncoder()
+    encoding.writeVarUint(encoder, messageSync)
+    const syncMessageType = syncProtocol.readSyncMessage(decoder, encoder, this.doc, this)
+    if (syncMessageType === syncProtocol.messageYjsSyncStep2) {
+      this.setSynced(true)
+    }
+
+    if (encoding.length(encoder) > 1) {
+      this.websocket?.send(encoding.toUint8Array(encoder))
     }
   }
 
-  disconnect() {
-    this.shouldConnect = false
-    this.disconnectBc()
-    if (this.ws !== null) {
-      this.ws.close()
+  private websocketOpen() {
+    this.setStatus({ status: STATUS_CONNECTED })
+    this.syncStep1()
+    this.broadcastAwareness()
+  }
+
+  private receiveMessage(event: MessageEvent) {
+    let message: Uint8Array = new Uint8Array(event.data)
+    const decoder = decoding.createDecoder(message)
+    const messageType = decoding.readVarUint(decoder)
+    switch (messageType) {
+      case messageSync:
+        this.receiveSyncMessage(decoder)
+        break
+      case messageAwareness:
+        this.updateAwareness(decoder)
+        break
+      case messageQueryAwareness:
+        this.queryAwareness()
+        break
+      default:
+        break
     }
   }
 
-  connect() {
-    this.shouldConnect = true
-    if (!this.wsconnected && this.ws === null) {
-      setupWS(this)
-      this.connectBc()
+  private websocketClose(event: CloseEvent) {
+    this.emit('connection-close', event)
+    this.setSynced(false)
+    this.setStatus({ status: STATUS_DISCONNECTED })
+
+    // Remove all awareness states except for our own.
+    awarenessProtocol.removeAwarenessStates(
+      this.awareness,
+      Array.from(this.awareness.getStates().keys()).filter(
+        (client) => client !== this.doc.clientID,
+      ),
+      this,
+    )
+  }
+
+  private broadcastAwareness() {
+    if (this.awareness.getLocalState() !== null) {
+      const encoderAwarenessState = encoding.createEncoder()
+      encoding.writeVarUint(encoderAwarenessState, messageAwareness)
+      encoding.writeVarUint8Array(
+        encoderAwarenessState,
+        awarenessProtocol.encodeAwarenessUpdate(this.awareness, [this.doc.clientID]),
+      )
+      this.websocket?.send(encoding.toUint8Array(encoderAwarenessState))
     }
   }
 
-  addOnFailureHandler(handler: () => void): () => void {
-    this.onFailureHandlers.push(handler)
-    return () => {
-      this.onFailureHandlers = this.onFailureHandlers.filter((h) => h !== handler)
+  private websocketError(event: Event) {
+    this.emit('connection-error', event)
+  }
+
+  protected emit(eventName: YSweetEvent, data: any = null): void {
+    const listeners = this.listeners.get(eventName)
```

Comment:
not a huge issue but we can avoid a level of nesting by falling back to an empty set
```suggestion
    const listeners = this.listeners.get(eventName) || new Set()
```

Reply:
Opted not to, I think this makes sense if we are getting a set to read, but since we are getting a set to write I think it actually obfuscates the purpose of the code when we create a value just to do a throwaway mutation to it.

Reply:
Isn't this a read? We're just iterating through the members and calling them. Regardless nbd, feel free to leave as is

Reply:
Isn't it a read? We're just iterating through the members and calling them. Regardless nbd, feel free to leave as is

Reply:
ooh sorry I did a ctrl+f in the source and ended up in the wrong place. Yep, that's just a read, I like your approach then.